apis/Google.Cloud.Metadata.V1/Google.Cloud.Metadata.V1/MetadataClientImpl.cs

```diff
@@ -0,0 +1,502 @@
+using Google.Apis.Compute.v1.Data;
+using Google.Apis.Http;
+using Google.Apis.Json;
+using Google.Apis.Logging;
+using Microsoft.CSharp.RuntimeBinder;
+using Newtonsoft.Json.Linq;
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Linq;
+using System.Net.Http;
+using System.Text.RegularExpressions;
+using System.Threading;
+using System.Threading.Tasks;
+using static Google.Apis.Compute.v1.Data.Metadata;
+
+namespace Google.Cloud.Metadata.V1
+{
+    /// <summary>
+    /// Wrapper around metadata server requests to provide simpler operations.
+    /// </summary>
+    /// <remarks>
+    /// This is the "default" implementation of <see cref="MetadataClient"/>. Most client code
+    /// should refer to <see cref="MetadataClient"/>, creating instances with
+    /// <see cref="MetadataClient.Create(ConfigurableHttpClient)"/> and
+    /// <see cref="MetadataClient.CreateAsync(IHttpClientFactory)"/>. The constructor
+    /// of this class is public for the sake of constructor-based dependency injection.
+    /// </remarks>
+    public sealed class MetadataClientImpl : MetadataClient
+    {
+        private const string DefaultMetadataHost = "metadata";
+        private const string EmulatorEnvironmentVariable = "METADATA_EMULATOR_HOST";
+        private const string MetadataFlavor = "Metadata-Flavor";
+        private const string GoogleMetadataHeader = "Google";
+
+        // TODO: Should this be created with a MetadataClient instead?
+        private static readonly ILogger Logger = ApplicationContext.Logger.ForType<MetadataClientImpl>();
+
+        private readonly string metadataHost;
+
+        private readonly object eventLock = new object();
+        private ValueWatcher instanceWatcher;
+        private ValueWatcher maintenanceStatusWatcher;
+        private ValueWatcher projectWatcher;
+
+        public override ConfigurableHttpClient HttpClient { get; }
+
+        /// <summary>
+        /// Constructs a new client for the metadata server.
+        /// </summary>
+        /// <param name="httpClient">The <see cref="ConfigurableHttpClient"/> to use when making requests, or null to create a default instance.</param>
+        public MetadataClientImpl(ConfigurableHttpClient httpClient = null)
+        {
+            HttpClient = httpClient ?? new HttpClientFactory().CreateHttpClient(new CreateHttpClientArgs());
+            metadataHost = GetMetadataHost();
+        }
+
+        /// <summary>
+        /// Constructs a new client for the metadata server.
+        /// </summary>
+        /// <param name="httpClientFactory">
+        /// An <see cref="IHttpClientFactory"/> used to create a <see cref="ConfigurableHttpClient"/> for making requests, or null to
+        /// create a default instance.
+        /// </param>
+        public MetadataClientImpl(IHttpClientFactory httpClientFactory)
+            : this((httpClientFactory ?? new HttpClientFactory()).CreateHttpClient(new CreateHttpClientArgs())) { }
```

Comment:
Same as before, can we split this up somehow?

Small nit: move the `{}` to the new line.